### PR TITLE
Major refactor: Introduce ODP in new structure (service,api,repository)

### PR DESCRIPTION
**Pull request by sorja**
Note:

Agreed to have ODP (Original Data Point) over NDP (National Data Point)

Further occurrences of NDP will be changed to ODP

### CODE REVIEW

File: server/eof/api.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import * as R from 'ramda'
 
 import { ApiAuthMiddleware } from '@server/api/middleware'
+import { readEofOdps } from '@server/repository/odp/readEofOdps'
+import { readFocOdps } from '@server/repository/odp/readFocOdps'
```
Comment (minotogna) [inline server/eof/api.ts:5]: why not to use OdpService?

File: server/eof/fraValueService.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { getDynamicCountryConfiguration } from '@server/repository/country/getDynamicCountryConfiguration'
+import { readEofOdps } from '@server/repository/odp/readEofOdps'
+import { readFocOdps } from '@server/repository/odp/readFocOdps'
```
Comment (minotogna) [inline server/eof/fraValueService.ts:3]: we need to think about FraValueService. Mayne we should use OdpService in this case?

File: server/repository/odp/deleteDraft.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  const actualId = actualRes.rows[0].actual_id
+  if (actualId) {
+    await client.query('UPDATE odp SET draft_id = null WHERE id = $1', [odpId])
+    const odpVersionId = await OdpService.getOdpVersionId(client, odpId)
+    const odpClasses = await getOdpNationalClasses(client, odpVersionId)
+    return wipeNationalClassIssues(client, odpId, countryIso, odpClasses)
+  }
+
+  return OdpService.deleteOdp(client, odpId, user)
+}
```
Comment (minotogna) [inline server/repository/odp/deleteDraft.ts:17]: repository cannot use service.

plus these operations should be executed in transactions created in Service.
should we talk about this?

File: server/repository/odp/listOriginalDataPoints.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import * as db from '@server/db/db'
+import * as Promise from 'bluebird'
```
Comment (minotogna) [inline server/repository/odp/listOriginalDataPoints.ts:2]: not sure if we should still use Bluebird. what do you think?

File: server/repository/odp/saveDraft.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { updateOrInsertDraft } from '@server/repository/odpDraft/odpDraftRepository'
+import { OdpService } from '@server/service'
+
+export const saveDraft = async (client: any, countryIso: any, user: any, draft: any) => {
+  const odpId = draft.odpId ? draft.odpId : await OdpService.createOdp(client, countryIso, user)
```
Comment (minotogna) [inline server/repository/odp/saveDraft.ts:5]: never use service in repository

File: server/service/assessment/ndp/ndpExporter.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ export const normalizeValue = R.pipe(
 )
 
 export const getCountryData = async (country: any) => {
-  const dataPoints = await NdpRepository.listOriginalDataPoints(country.countryIso)
+  const dataPoints = await OdpService.listOriginalDataPoints(country.countryIso)
```
Comment (minotogna) [inline server/service/assessment/ndp/ndpExporter.ts:26]: this exporter should be moved under service?

File: server/repository/odp/listOriginalDataPoints.ts (hunk 4 above)
Comment (sorja) [reply server/repository/odp/listOriginalDataPoints.ts:2]: We can remove, but in this file it was not straightforward to replace iirc
(Would prefer removing)

File: server/eof/api.ts (hunk 1 above)
Comment (sorja) [reply server/eof/api.ts:5]: This file needs refactoring anyway (still part of apiRouter), fixing for now

File: server/repository/odp/listOriginalDataPoints.ts (hunk 4 above)
Comment (minotogna) [reply server/repository/odp/listOriginalDataPoints.ts:2]: please, remove it. future refactor will remove it as well

File: server/repository/odp/deleteDraft.ts (hunk 3 above)
Comment (minotogna) [reply server/repository/odp/deleteDraft.ts]: just don't use service in repository for now.
will make a new task for transactions

File: server/repository/odp/deleteDraft.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  const actualRes = await client.query('SELECT actual_id FROM odp WHERE id = $1', [odpId])
+  const actualId = actualRes.rows[0].actual_id
+  if (actualId) {
+    await client.query('UPDATE odp SET draft_id = null WHERE id = $1', [odpId])
+    const odpVersionId = await OdpRepository.getOdpVersionId(client, odpId)
+    const odpClasses = await getOdpNationalClasses(client, odpVersionId)
+    return wipeNationalClassIssues(client, odpId, countryIso, odpClasses)
+  }
+
+  return OdpRepository.deleteOdp(client, odpId, user)
```
Comment (minotogna) [inline server/repository/odp/deleteDraft.ts:16]: don't you have circular dependency in this case?
cant you import deleteOdp from './deleteOdp' ?

File: server/repository/odp/deleteOdp.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import * as Promise from 'bluebird'
+import { wipeClassData } from '@server/repository/odpClass/odpClassRepository'
+import { deleteIssues } from '@server/repository/review/reviewRepository'
+import { insertAudit } from '@server/repository/audit/auditRepository'
+import { OdpRepository } from '@server/repository'
+import { getAndCheckOdpCountryId } from './getAndCheckOdpCountryId'
+
+export const deleteOdp = async (client: any, odpId: any, user: any) => {
+  const countryIso = await getAndCheckOdpCountryId(client, odpId, user)
+  const odpVersionId = await OdpRepository.getOdpVersionId(client, odpId)
```
Comment (minotogna) [inline server/repository/odp/deleteOdp.ts:10]: same as above

File: server/repository/odp/getOdp.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import * as camelize from 'camelize'
+
+import * as db from '@server/db/db'
+import { getOdpNationalClasses } from '@server/repository/odpClass/odpClassRepository'
+import { OdpRepository } from '@server/repository'
+
+export const getOdp = async (odpId: any, schemaName = 'public') => {
+  const versionId = await OdpRepository.getOdpVersionId(db.pool, odpId, schemaName)
```
Comment (minotogna) [inline server/repository/odp/getOdp.ts:10]: same as above

File: server/repository/odp/deleteDraft.ts (hunk 7 above)
Comment (sorja) [reply server/repository/odp/deleteDraft.ts:16]: automatic import, can fix

File: server/service/assessment/ndp/ndpExporter.ts (hunk 6 above)
Comment (sorja) [reply server/service/assessment/ndp/ndpExporter.ts:26]: separate issue

**Review by minotogna — APPROVED** [review]

**Pull request merged**